### PR TITLE
ARROW-5131: [Python] Support Azure Data Lake Filesystem

### PR DESCRIPTION
**Pull request by hayesgb**
This PR adds a Gen1 Azure Datalake Filesystem Wrapper, in filesystem.py, which enables the ability read parquet files written to Azure Datalake using the pyarrow engine.  To work, this also requires explicitly importing the dask-adlfs package, as found [here](https://github.com/dask/dask-adlfs).

### CODE REVIEW

File: python/pyarrow/parquet.py [hunk 1]
```diff
@@ -1044,6 +1044,8 @@ def _make_manifest(path_or_paths, fs, pathsep='/', metadata_nthreads=1):
         path_or_paths = path_or_paths[0]
 
     if _is_path_like(path_or_paths) and fs.isdir(path_or_paths):
+        print('...This is a directory... checking fs.pathsep')
```
Comment (emkornfield) [inline python/pyarrow/parquet.py]: pleaes remove.

File: python/pyarrow/filesystem.py [hunk 2]
```diff
@@ -371,9 +371,81 @@ def walk(self, path, refresh=False):
                 yield tup
 
 
-def _sanitize_s3(path):
+class ADLFSWrapper(DaskFileSystem):
+
```
Comment (emkornfield) [inline python/pyarrow/filesystem.py]: Please document.

File: python/pyarrow/filesystem.py [hunk 3]
```diff
@@ -371,9 +371,81 @@ def walk(self, path, refresh=False):
                 yield tup
 
 
-def _sanitize_s3(path):
+class ADLFSWrapper(DaskFileSystem):
```
Comment (emkornfield) [inline python/pyarrow/filesystem.py]: nit: I think I would prefer to spell this out, (At lease Azure)

File: python/pyarrow/filesystem.py [hunk 4]
```diff
@@ -371,9 +371,81 @@ def walk(self, path, refresh=False):
...
+        for key in list(self.fs._ls(path, invalidate_cache=invalidate_cache)):
+            path = key['name']
+            if key['type'] == 'DIRECTORY':
+                directories.add(path)
+            elif key['type'] == 'FILE':
+                pass
+            else:
+                files.add(path)
+
+        # s3fs creates duplicate 'DIRECTORY' entries
```
Comment (emkornfield) [inline python/pyarrow/filesystem.py]: copy and paste bug?

File: python/pyarrow/filesystem.py [hunk 5]
```diff
@@ -371,9 +371,81 @@ def walk(self, path, refresh=False):
...
+                yield tup
+
+
+def _sanitize_remote_path(path):
     if path.startswith('s3://'):
         return path.replace('s3://', '')
+    elif path.startswith('adl://'):
+        newpath = path.replace('adl://', '')
+        path_parts = newpath.split('/')
+        path_parts = [p for p in path_parts if 'azuredatalakestore.net' not
```
Comment (emkornfield) [inline python/pyarrow/filesystem.py]: please comment your intent here

File: python/pip-wheel-metadata/pyarrow.dist-info/LICENSE.txt [hunk 6]
```diff
@@ -0,0 +1,88 @@
+## 3rd-party licenses for code that has been adapted for the Arrow Python
```
Comment (emkornfield) [inline python/pip-wheel-metadata/pyarrow.dist-info/LICENSE.txt]: is this change required for this PR?  If not maybe add it as a separate JIRA/PR

**Review by emkornfield — CHANGES_REQUESTED** [review]
Hi @hayesgb thank you for the PR.  The convention for naming PRs is to start with the JIRA, then provide a language tage and descrption (i.e. ARROW-5131: [Python] Add Azure Datalake filesystem.

I posted some initial feedback, I will try to take another pass in a day or two.  Would it be possible to add some unit tests to the PR?

File: python/pyarrow/parquet.py (hunk 1 above)
Comment (hayesgb) [reply python/pyarrow/parquet.py]: Removed in next push.

File: python/pyarrow/filesystem.py (hunk 2 above)
Comment (hayesgb) [reply python/pyarrow/filesystem.py]: Added to upcoming PR

File: python/pyarrow/filesystem.py (hunk 3 above)
Comment (hayesgb) [reply python/pyarrow/filesystem.py]: Changed to AzureDatalakeFilesystemWrapper.  This will require a change to the [dask-adlfs](https://github.com/hayesgb/dask-adlfs/blob/master/dask_adlfs/core.py), which I will create and submit once this is complete.

File: python/pyarrow/filesystem.py (hunk 4 above)
Comment (hayesgb) [reply python/pyarrow/filesystem.py]: Yes.  Comment removed.

File: python/pyarrow/filesystem.py (hunk 5 above)
Comment (hayesgb) [reply python/pyarrow/filesystem.py]: Added documentation

File: python/pip-wheel-metadata/pyarrow.dist-info/LICENSE.txt (hunk 6 above)
Comment (hayesgb) [reply python/pip-wheel-metadata/pyarrow.dist-info/LICENSE.txt]: No.  It's removed.

File: python/pip-wheel-metadata/pyarrow.dist-info/LICENSE.txt [hunk 7]
```diff
@@ -0,0 +1,87 @@
+
```
Comment (emkornfield) [inline python/pip-wheel-metadata/pyarrow.dist-info/LICENSE.txt]: sorry in my previous comment I was asking generally if the pip-wheel stuff is required in this PR or is the python code enough (I'm don't have any experience with wheel packaging).

**Comment by emkornfield** [comment]
Sorry if I missed a response, but is there any way to add unit/integration tests to this code.  (I'm not sure what we do for s3, @pitrou @kszucs?

File: python/pip-wheel-metadata/pyarrow.dist-info/LICENSE.txt (hunk 7 above)
Comment (emkornfield) [reply python/pip-wheel-metadata/pyarrow.dist-info/LICENSE.txt]: If they are required you have to add an apache header or exclude them, so CI passes.

**Comment by pitrou** [comment]
@emkornfield I don't think so, unless there's a way to mock the filesystem.

File: python/pip-wheel-metadata/pyarrow.dist-info/LICENSE.txt (hunk 7 above)
Comment (pitrou) [inline python/pip-wheel-metadata/pyarrow.dist-info/LICENSE.txt]: The whole `pip-wheel-metadata` directory looks like it shouldn't be in this PR, IMHO.

File: python/pyarrow/filesystem.py [hunk 8]
```diff
@@ -371,9 +371,99 @@ def walk(self, path, refresh=False):
...
+                              for x in directories])
+
+        yield path, directories, files
+
+        for directory in directories:
+            for tup in self.walk(directory, refresh=refresh):
+                yield tup
+
+
+def _sanitize_remote_path(path):
```
Comment (pitrou) [inline python/pyarrow/filesystem.py]: Hmm, we should instead have a specific function for each filesystem implementation. So, beside `_sanitize_s3`, add `_sanitize_adl`. Also, can you add unit tests for this in `test_filesystem.py`?

File: python/pyarrow/filesystem.py [hunk 9]
```diff
@@ -371,9 +371,99 @@ def walk(self, path, refresh=False):
...
+    @implements(FileSystem.isfile)
+    def isfile(self, path):
+        path = _sanitize_remote_path(_stringify_path(path))
+        try:
+            contents = self.fs.ls(path)
+            return len(contents) == 1 and contents[0] == path
+        except OSError:
+            return False
+
+    def walk(self, path, invalidate_cache=True, refresh=False):
```
Comment (pitrou) [inline python/pyarrow/filesystem.py]: Why are there distinct options `refresh` and `invalidate_cache`?

File: python/pyarrow/filesystem.py [hunk 10]
```diff
@@ -371,9 +371,99 @@ def walk(self, path, refresh=False):
...
+        files = set()
+
+        for key in list(self.fs._ls(path, invalidate_cache=invalidate_cache)):
+            path = key['name']
+            if key['type'] == 'DIRECTORY':
+                directories.add(path)
+            elif key['type'] == 'FILE':
+                pass
+            else:
+                files.add(path)
```
Comment (pitrou) [inline python/pyarrow/filesystem.py]: I don't understand this. Nothing happens if the type is 'FILE'?

File: python/pyarrow/filesystem.py [hunk 11]
```diff
@@ -371,9 +371,99 @@ def walk(self, path, refresh=False):
...
+
+        files = sorted([posixpath.split(f)[1] for f in files
```
Comment (pitrou) [inline python/pyarrow/filesystem.py]: `posixpath.basename(f)` would be easier to understand, IMHO.

File: python/pyarrow/filesystem.py [hunk 12]
```diff
@@ -371,9 +371,99 @@ def walk(self, path, refresh=False):
...
+                        if f not in directories])
```
Comment (pitrou) [inline python/pyarrow/filesystem.py]: I don't understand this. A file name can be both in `files` and `directories`?

File: python/pyarrow/filesystem.py [hunk 13]
```diff
@@ -371,9 +371,99 @@ def walk(self, path, refresh=False):
...
+    @implements(FileSystem.isdir)
+    def isdir(self, path):
+        path = _sanitize_remote_path(_stringify_path(path))
+        try:
+            contents = self.fs.ls(path)
+            if len(contents) == 1 and contents[0] == path:
+                return False
+            else:
+                if not any(ftype in path for ftype in
+                           ['parquet', 'parq', 'metadata']):
```
Comment (pitrou) [inline python/pyarrow/filesystem.py]: I don't think this has anything to do with this PR.

File: python/pyarrow/filesystem.py [hunk 14]
```diff
@@ -371,9 +371,99 @@ def walk(self, path, refresh=False):
                 yield tup
 
 
-def _sanitize_s3(path):
+class AzureDatalakeFileSystemWrapper(DaskFileSystem):
+    """
+    Wraps the Gen1 Azure Datalake Filesystem client.
```
Comment (pitrou) [inline python/pyarrow/filesystem.py]: Can you add a short example of how this is supposed to be used? And mention the `dask-adl` package?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for posting this PR! Here are some questions and comments.

**Comment by pitrou** [comment]
@martindurant what is the state of dask-adlfs project? Do you think it can be relied on?

**Comment by martindurant** [comment]
It is not particularly watched by anyone - but the code is very simple, so the chances of things going wrong are relatively small. There are two things I would mention:
- with my work on [fsspec](https://github.com/martindurant/filesystem_spec), I intend to rewrite adlfs along with other file-systems, so that they all share the same framework code and will automatically become compatible with dask and arrow
- azure has released "generation 2" of their API, joining datalake and blob, so adlfs will presumably be retired at some point. I don't know the timescale for this.

**Comment by wesm** [comment]
Along with S3 and GCS I suggest we plan to develop a C++ interface to Azure Blob Storage so we can utilize a common codebase for interacting with datasets from Python, R, Ruby, and others. Otherwise we could end up playing bug whackamole across different programming languages

References

* S3 https://issues.apache.org/jira/browse/ARROW-453
* Google https://issues.apache.org/jira/browse/ARROW-1231
* Azure https://issues.apache.org/jira/browse/ARROW-2034

**Comment by martindurant** [comment]
You may be interested in my work to unify the file-systems on the python side: https://github.com/martindurant/filesystem_spec , if only for (an) inspiration.

Note that Azure have introduced an "API 2" to succeed from both blob and datalake - I haven't yet looked into whether easily usable.

**Comment by martindurant** [comment]
(of course this was mentioned before, just noting it again - also that any filesystem defined within fsspec is automatically compatible with *py*arrow)

**Comment by wesm** [comment]
@martindurant yes, we will definitely look at what you've done (and at other prior art, e.g. TensorFlow's filesystem stuff) to inform the C++ API. A lot of the requirements will be driven by the downstream use in the consumer languages. The filesystem implementations can be exposed in Python with a common API

**Comment by emkornfield** [comment]
@pitrou @wesm what did you want to do with this PR?  @hayesgb do you have bandwidth to address the comments?

**Comment by pitrou** [comment]
@emkornfield As for me, nothing special.

**Comment by hayesgb** [comment]
In looking at the Azure Datalake Gen2 filesystem spec, they've written the "abfs://" protocol into the hdfs specification, so in principle it should be incorporated into that specification.  See [here](https://hadoop.apache.org/docs/current/hadoop-azure/index.html).  In principle, the abfs files hould be available via hdfs3 and/or libhdfs3, but I'm on Windows and those packages are only available for linux.  Can anyone comment on that?

I like the idea of writing a cpp filesystem implementation, but realistically it will be a while before I can take a run at that.  I'm more likely to get to the fsspec suggestion first.  Additionally, I found the following [project](https://github.com/manish/dask-azureblobfs), which I would be willing to see how easily it could be incorporated, either by reaching out to the author or forking the work.

**Comment by emkornfield** [comment]
Is there a clear path forward here?  It seems like progress has stalled?

**Comment by hayesgb** [comment]
In light of the existing dependency, it makes perfect sense.  I can take a
look at rewriting it using fsspec.  Azure is also making progress toward a
Python SDK for their Datalake Gen2, as documented here
<https://azure.microsoft.com/en-us/blog/silo-busting-2-0-multi-protocol-access-for-azure-data-lake-storage/>
.

**Comment by emkornfield** [comment]
@hayesgb could we maybe close this PR and reopen a new one when fsspec is ready?

**Comment by hayesgb** [comment]
Makes sense.  I’m working on a implementation of both the ADLFS Gen1 and Gen2 filesystems that leverage fsspec.

**Comment by pitrou** [comment]
Note that the plan for Arrow is to write all filesystem implementations in C++, in order to make them available to many implementations and not require a Python runtime when doing C++-based data processing.

The `master` branch has an abstract filesystem API as well as "local filesystem" and "mock" (in-memory) implementations of that API:
https://github.com/apache/arrow/tree/master/cpp/src/arrow/filesystem

I'm currently working on an S3 implementation and hope to have a draft available in circa one week. Hopefully other contributors will want to tackle other widely-used cloud storage APIs so that Arrow gets a comprehensive set of filesystem implementations.

**Comment by hayesgb** [comment]
Thanks for sharing the work.  I would be curious to understand how / if there’s a relationship between this work and the work of that’s being done on the fsspec by martindurant (referenced earlier in this thread).  It looks like fsspec has become a dependency for Dask.  As of today, there’s a DaskFileSystem class in pyarrow.  Do you see this living on, but with a C++ Filesystem base class, or do you see C++ filesystem implementations for all major cloud vendors?

**Comment by pitrou** [comment]
For now those aren't related. At some point, perhaps it will be possible to make a `fsspec`-compliant wrapper around Arrow filesystem implementations. But our aim is provide C++ implementations for all important storage APIs.

**Comment by wesm** [comment]
Yes, to add some context around this, we are members of our community that are keen to have identical dataset/filesystem functionality in R and Ruby (and probably more languages eventually) in addition to Python, so we need core functionality to be independent from Python. So any development work that's happening in pure Python won't be relevant from an implementation point of view except to help communicate application / functional requirements.

**Comment by emkornfield** [comment]
Closing this it seems like there is potentially more follow-up work once the file system APIs come online.
